### DISCUSSION TITLE
MCA counter example

### DISCUSSION THREAD

**Discussion by ddavid42** [verificarlo/verificarlo#347] [General]
While conducting a benchmark, I encountered an interesting example where MCA, RR, and PB struggled to provide an accurate analysis (or at least made it challenging to demonstrate a complete loss of accuracy).

Here’s a straightforward example. I deliberately kept the value of N small, but even when increasing it to N=128, Verificarlo fails to deliver useful insights.
 
==============
```C
  #include <math.h>
  #include <stdio.h>
  #include <stdlib.h>
  
  double square_root(double x, int N){
      double res=x;
  
      for(int i=0; i<N; i++){
          res = sqrt(res);
      }
  
      return res;
  }
  
  double square(double x, int N){
      double res=x;
  
      for(int i=0; i<N; i++){
          res = res*res;
      }
  
      return res;
  }
  
  
  int main(int argc, char **argv){
      for(int N = 10; N<80; N+=10){
          for(double x=0.0; x<4.0; x+=0.5){
              double r = square_root(x, N);
              double s = square(r, N);
              printf("x=%f, N=%d, S=square_root(x,N)=%e, square(s,N)=%e\n", x, N, r, s);
          }
          printf("\n");
      }
      return 0;
  }
```

**Comment by pablooliveira** [comment]
Dear @ddavid42,

Thanks for the benchmark. Could you please explain clearly what is that you are missing?

I tested your code with the `libinterflop_mca.so` and default parameters. Ideally, I would do this with 30 samples at least, but just to demonstrate I take three MCA samples.

I compute the number of significant digits with $s = -\log(\sigma/|\mu|)$. Also ideally, one would apply a confidence interval, but here it is not necessary.

for N=60, x=4.0, I get the following three output samples 2.694926e+07, 1.638478e-28, 6.709994e+09
=> s ~ -0.34 < 0 (no significant digits at all, complete loss of significance)

for N=70, x=4.0, I get 0.000000e+00, inf, 0.000000e+00
=> s < 0 (no significant digits at all, complete loss of significance)

So, according to my brief tests, MCA does indeed report a total loss of significance here. 

If I did miss something, could you please explain what it is that you expect to see?
If this answers your question, then please tell me so I can close this issue.

thanks!

**Comment by ddavid42** [comment]
Indeed, this is not a bug but rather a concept which is difficult to capture with MCA. 

This concept is related to the concept of information's vanishing (which is different from cancellation).
By iteratively computing x_{n+1}=sqrt(x_{n}), every perturbation (on the input or the output) will slowly vanishes along with the true informations in order to produce confidently 1.0 at the end with 100% probability. 

However, this program is behaving as if MCA computation were done internally in extended precision:
for N=60, eventhough S=square_root(2.0, 60)=1.0000.... exactly, square(2.0,60) = "some observables variations".
The same behaviour is observable untill N=63, and disapear for N>65, meaning  S=square_root(2.0, N>65)=1.0000.... exactly, square(2.0, N>65) = 0.0 (with high probability).

Second observation: 
In MCA, for very large value of N, square(1.0 , N) = Inf (with low probability), which is another issues with MCA as it could lead the users to think that during his computations, he lost everything (which is not the case)... concept of noise amplification.
 
My purpose was just to point out some limitations of MCA/Verificarlo, and this issues could be closed. 
Thanks you again for this great tool.

**Comment by pablooliveira** [comment]
Dear @ddavid42,

Some clarifications : 
- Square root is not instrumented with MCA in any case. Currently, verificarlo does not instrument sqrt. (We can do it using libfuncinstrument + vprec backend, but note we do not define MCA for square roots).

- Indeed, MCA will not deterministically converge to 1.0. This is interesting, but not specific to this benchmark; there are different cases where MCA will not converge to the IEEE-754 computed value.

- MCA computation within Verificarlo is an approximation of MCA done using limited precision (see https://tel.archives-ouvertes.fr/tel-03831483/); higher precision is possible but would increase the instrumentation cost, so we need to have a tradeoff so the tool can be used on large codebases.

- MCA does though show a complete loss of accuracy.

Making this a conversation, since it is not an issue or a bug.

Thanks for the report.

**Comment by yohanchatelain** [comment]
Hello @ddavid42,

Thank you for providing the example! However, isn’t this a case similar to Muller’s sequence? Where you lose precision over time but finally get a stable results at the end?

For me it is also a great example of why using tracers like VeriTracer or PyTracer to track floating-point behaviour in detail, making it easier to identify where and how instability arises in numerical computations.

**Comment by ddavid42** [comment]
Hello  @yohanchatelain ,

The first part of this example iterative call to sqrt() is characteristic of progressive loss of significance and differs from Muller's sequence which exhibits 2 attractors.

In details, it's seems that this is actually the PB mode which exhibit a bias toward 0 at the end. 
For the squaring function in PB mode with a large value of N, we expect the function to either converge toward 0 or inf with a probability of 50%. However, this is not the case as the probability of returning 0 appears to be ~70%.
Here is a simplified example running in PB mode:

```C
#include <stdio.h>

double square(double x, int N){
    double res=x;

    for(int i=0; i<N; i++){
        res = res * res;
    }
    return res;
}

int main(int argc, char **argv){
    int z=0;
    int nbiter = 1000;

    for(int l=0; l<nbiter; l++){    
        double s = square(1.0, 80);
        printf("square(1.0, 80)=%.20a\n", s);
        if (s==0) z++;
    }
        
    printf("Probability of 0.0 z = %.2f %%\n", (float)z/nbiter*100);
    return 0;
}
```

**Reply by pablooliveira** [reply]
Dear @ddavid42, 

Thanks for the analysis.

PB mode or full-MCA mode are indeed biased, this is a known property and not a bug. They are, nevertheless, very useful in highlighting cancellations.

Only RR mode (also called stochastic rounding) is unbiased for basic operations. This morning you asked me in our discussion a reference, you can find Stott Parker's proof and discussion in section 6.4 of his manuscript (https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=3b55cfb2325bc081f113612204733c2a3bd650fe).

To really have an unbiased RR operator, Stott Parker's definition of RR requires a correction; I discuss that correction and prove unbiasedness in section 5.4 of https://tel.archives-ouvertes.fr/tel-03831483/file/habilitation.pdf.

**Reply by pablooliveira** [reply]
To be complete I have tried your code on my side:
- for RR, I always get the correct result 1.0, so the result is unbiased. (Operations are exact for 2⁰).
- for MCA, I have z = 43% (biased)
- for PB, though I cannot reproduce your results, I seem to have z ~ 50% (sometimes over, sometimes under); so our  results seem to differ slightly here.

Not sure where this difference is coming, maybe different compiler options ?

**Reply by yohanchatelain** [reply]
The main difference between the PB implementation in Verificarlo and the one developed in MCA is whether the number is representable at the selected virtual precision. In MCA theory, Parker introduced the function `t_digit_precision(x)`, which equals `x` if the number is exactly representable with `t` digits, or `inexact(x)` otherwise. While it's possible to verify this in the RR mode (since intermediate calculations are done with at least `2 * p` digits, where `p` is the precision of the floating-point number), it’s not feasible for PB in its current implementation since it would require saving results at a higher precision.

Additionally, I tried reproducing your code and observed the following:
- z = 70% when using `libinterflop_mca.so --mode=pb`
- z = 50% when using `libinterflop_mca_int.so --mode=pb`

This might explain the difference in your results.

**Reply by ddavid42** [reply]
You pinpoint our difference and I'm able to get 50% with libinterflop_mca_int.so as well now,...
So  libinterflop_mca_int.so exhibit the expected behaviour but not libinterflop_mca.so, hum !

**Reply by pablooliveira** [reply]
I agree that PB does not preserve exact computations. But even though, PB/FULLMCA are not unbiased, so distributions of results that are centered around the exact value should not be expected in general (with PB/FULLMCA) !

Regarding the difference, you nailed it @yohanchatelain, indeed the difference of z comes from the backend. For the most exact RR, `mca_int.so` is required and faster.
`mca.so` is interesting when t != p.

**Reply by pablooliveira** [reply]
> @ddavid42 hum !
 
 Indeed, mca_int implements the rounding operator correction I mentioned above (section 5.4). For strict unbiased RR (SR) it is the preferred backend.

**Comment by pablooliveira** [comment]
These discussions highlight the need to guide the user's choice for a MCA backend.
I have proposed a documentation improvement in https://github.com/verificarlo/verificarlo/pull/348.
